### PR TITLE
Refactoring of map alignment code

### PR DESCRIPTION
**Pull request by hendrikweisser**
Fixes problems listed in #1523 and adds a feature (storing of original RT values) needed for #1500.

In more detail:
- storing of original RTs was added to class MapAlignmentTransformer and exposed in MapRTTransformer (TOPP tool)
- base class MapAlignmentAlgorithm was removed
- handling of references in MapAlignmentAlgorithmIdentification was improved (no more reading of input files in the library class)
- non-shared code from TOPPMapAlignerBase was moved to MapAligner classes
- lots of minor clean-ups
- update to pyOpenMS (not fully tested, currently doesn't build due to #1527)

I considered whether it would make sense to remove the class TOPPMapAlignerBase (in "src/openms/include/OpenMS/APPLICATIONS/MapAlignerBase.h") altogether, but there is a lot of code for registering and consistency checking of parameters that is shared between the three MapAligner tools, that would have to be duplicated - so I think it's better to keep this class for now.

### CODE REVIEW

File: src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/MapAlignmentAlgorithmIdentification.h [hunk 1]
```diff
@@ -74,79 +81,63 @@ namespace OpenMS
...
+                                         __PRETTY_FUNCTION__, reference_index, 
+                                         data.size());
+        }
+        setReference(data[reference_index]);
       }
 
       // one set of RT data for each input map, except reference:
-      std::vector<SeqToList> rt_data(maps.size() - bool(reference_index_));
-      for (Size i = 0, j = 0; i < maps.size(); ++i)
+      std::vector<SeqToList> rt_data(data.size() - (reference_index >= 0));
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/MapAlignmentAlgorithmIdentification.h]: ```- (reference_index >= 0)``` is ugly and makes the code imho harder to read

Comment (hendrikweisser) [inline src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/MapAlignmentAlgorithmIdentification.h]: Well, there's a helpful comment above that line...
I can add a `bool use_reference = (reference_index >= 0);`, but I would still subtract the bool to get the correct size. The alternative would be to add an "if" clause and call `rt_data.resize` with two different values, but that seems unnecessarily long-winded.

**Comment by timosachsenberg** [comment]
Thanks alot for refactoring the MapAligner code base.
I think it is cleaner now and a clear improvement. I think we can continue with the merge assuming that changing the way references are counted is properly tested.

**Comment by hendrikweisser** [comment]
> I think it is cleaner now and a clear improvement.

Yes, thank you.

> I think we can continue with the merge assuming that changing the way references are counted is properly tested.

There is test coverage for that, so I'm quite sure it's fine. Travis also runs through (except for the usual "TOPP_PeakPickerWavelet_3" on one platform). Should be good to merge.

**Comment by timosachsenberg** [comment]
+1

**Comment by jpfeuffer** [comment]
I think these changes introduced some build errors on windows
http://cdash.openms.de/viewBuildError.php?buildid=139687

**Comment by hroest** [comment]
yes, I agree. I tried to address them but couldnt really figure out what Windows was complaining about. It seems that it is unhappy about `MetaInfoInterface& interface` but the error does not make sense to me

**Comment by hendrikweisser** [comment]
Sorry that there are problems with my pull request, but I'm in the same boat as Hannes: I can't see what the actual problem is and can't debug on Windows. @cbielow, could you have a look, please?

**Comment by jpfeuffer** [comment]
No problem. I looked on the build machines:

..\..\..\OpenMS\src\topp\MapAlignerPoseClustering.cpp(213): error C3016: 'i' : index variable in OpenMP 'for' statement must have signed integral type

Seems like that is the simple problem.

**Comment by lars20070** [comment]
@jpfeuffer That's the only correction necessary?
Then I will prepare a PR. I need the Windows builds for an update on our server.

**Comment by jpfeuffer** [comment]
I think a revert of the change from Int to Size in lines 213ff in MapAlignerPoseClustering.cpp should suffice.

**Comment by lars20070** [comment]
MapAlignerPoseClustering still writes a list of trafos as output. But MapRTTransformer now takes only a single input and single trafo. So -trafo_out is a list of files and -trafo_in is a single file. How are the two tools supposed to work together after #1531? There is no `split` node (reverse-collect) in TOPPAS.

**Comment by hendrikweisser** [comment]
(Continued discussion from #1574.)

@lars20070:
> Please revert MapRTTransformer to file list.

I doubt that's the right fix in this situation. We seem to have the general problem in TOPPAS that we can't connect tools that generate output file lists (only the MapAligners?) with tools that take single input files (most tools) - e.g. MapAlignerPoseClustering followed by TextExporter won't work, either. So I think we should urgently add a general remedy for that.

@cbielow: Can you comment on this? Is there perhaps already some functionality in TOPPAS that I'm overlooking?

**Comment by cbielow** [comment]
There is currently no splitter/Node in TOPPAS and I'd not know how to 
work around it.
This has been on the todo/list for quite some time, but we have had a 
pressing use case // I guess we have one now.

Am 9/2/2015 um 12:07 PM schrieb hendrikweisser:

**Comment by lars20070** [comment]
A `split` node would resolve our issue and be the cleaner solution.

**Comment by VOOginga** [comment]
Hi Chris
I would highly appreciate if you could implement the SPLIT NODE issue as discussed earlier. We really need it for our current workflow which is basically on hold. Thanks

**Comment by hendrikweisser** [comment]
@VOOginga:
> We really need it for our current workflow which is basically on hold.

I know it's not ideal and we should add a splitter node a.s.a.p., but if you have data to analyze now, you can always split the TOPPAS workflow into two between the MapAligner and MapRTTransformer.

@cbielow:
Do you have time to implement the splitter or do you want me to have a look next week?

**Comment by VOOginga** [comment]
Hi Chris
Kindly have a look at it next week. I would greatly appreciate. Thanks

@VOOginga <https://github.com/VOOginga>:

We really need it for our current workflow which is basically on hold.

I know it's not ideal and we should add a splitter node a.s.a.p., but if
you have data to analyze now, you can always split the TOPPAS workflow into
two between the MapAligner and MapRTTransformer.

@cbielow <https://github.com/cbielow>:
Do you have time to implement the splitter or do you want me to have a look
next week?

—
Reply to this email directly or view it on GitHub
<https://github.com/OpenMS/OpenMS/pull/1531#issuecomment-137952660>.

**Comment by cbielow** [comment]
@hendrikweisser If you want to have a look, please do. I probably wont find time in the near future.

@VOOginga Apart from what Hendrik suggested (workflow split), I can think of: use the OpenMS version before your workflow broke, OR just reintroduce list capabilities into the tool (it should be ~10 lines of code).

**Comment by lars20070** [comment]
IMO, splitting the workflow is the best solution for now. As soon as we have the split node in TOPPAS, we can go back to a fully automated label-free workflow.

**Comment by lars20070** [comment]
How does Merge, Collect and Split work in Knime?

**Comment by timosachsenberg** [comment]
Hi Lars,
in KNIME you have the ZipLoopStart and End node.

see:
https://github.com/OpenMS/UMTutorials/blob/master/Handout/graphics/knime_setup/Minimal_FileInfoLoop.pdf

and:
```
Workflows are typically constructed to process a large number of files automatically.
As a simple example, consider you would like to gather this information for more than one file.
We will now modify the workflow to compute the same information on three different files and then write the output files to a folder.

\begin{itemize}
\item
We start from the previous workflow.
\item
First we need to replace our single input file with multiple files.
Therefore we add the \KNIMENODE{Input Files} node from the category \menu{Community Nodes > GenericKnimeNodes > IO}.
\item
To select the files we double-click on the \KNIMENODE{Input Files} node and click on \menu{Add}.
In the filesystem browser we select all three files from the directory \directory{OpenMS / tiny / }.
And close the dialog with \menu{Ok}.
\item
We now add two more nodes: the \KNIMENODE{ZipLoopStart} and the \KNIMENODE{ZipLoopEnd} node from the category \menu{Community Nodes > GenericKnimeNodes > Flow}. 
\item
Afterwards we connect the \KNIMENODE{Input Files} node to the first port of the \KNIMENODE{ZipLoopStart} node, the first port of the \KNIMENODE{ZipLoopStart} node to the \KNIMENODE{FileInfo} node, the first output port of the \KNIMENODE{FileInfo} node to the first input port of the \KNIMENODE{ZipLoopEnd} node, and the first output port of the \KNIMENODE{ZipLoopEnd} node to the \KNIMENODE{Output Folder} node (, not \KNIMENODE{Output File}).
The complete workflow is shown in \cref{fig:knime_minimal_loop}
\item
The workflow is already complete.
Simply execute the workflow and inspect the output as before.
\end{itemize}

In case you had trouble to understand what ZipLoopStart and ZipLoopEnd do - here is a brief explanation:
\begin{itemize}
\item
The  \KNIMENODE{Input Files} node passes a list of files to the \KNIMENODE{ZipLoopStart} node.
\item
The \KNIMENODE{ZipLoopStart} node takes the files as input, but passes the single files sequentially (that is: one after the other) to the next node. 
\item
The \KNIMENODE{ZipLoopEnd} collects the single files that arrive at its input port. After all files have been processed, the collected files are passed again as file list to the next node that follows.
\end{itemize}
```

**Comment by lars20070** [comment]
ZipLoopStart == Collet
ZiplLoopEnd == Split
??? == Merge

**Comment by aiche** [comment]
There is also a `FileMerger` node in KNIME

**Comment by lars20070** [comment]
Thanks @aiche. That's a Knime `FileMerger` node not our OpenMS `FileMerger` tool. Right?

**Comment by aiche** [comment]
The node is part of the GKN base package.

**Comment by hendrikweisser** [comment]
Issue for a Splitter node: #1591
Pull request introducing one: #1593

**Pull request merged**